include/libtorrent/io.hpp

```diff
@@ -80,11 +81,16 @@ namespace libtorrent
 				|| std::is_enum<In>::value>::type>
 		inline void write_impl(In data, OutIt& start)
 		{
+			using OutItT = typename std::iterator_traits<OutIt>::value_type;
+			using Byte = typename std::conditional<
+				std::is_same<OutItT, void>::value, char, OutItT>::type;
```

Comment:
I think I get it now. you want this function to accept a ``void*`` as the output iterator, so users don't have to cast. I'm thinking of defining a ``byte`` type [1] instead, and try to use that pervasively for generic memory buffers. Maybe it would make sense to *require* the ``value_type`` be ``byte`` instead (as the final goal).

[1] http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2016/p0298r0.pdf